File: 4th Semester/Design and programming environments/C-Sharp Client-Server/protobuf3/ProtoChatWorker.cs
using App.Protocol;
using AppModel;
using AppServices;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;

namespace protobuf3
{
	public class ProtoChatWorker : IAppObserver
	{
		private IAppServices server;
		private TcpClient connection;

		private NetworkStream stream;
		//private IFormatter formatter;
		private volatile bool connected;
		public ProtoChatWorker(IAppServices server, TcpClient connection)
		{
			this.server = server;
			this.connection = connection;
			try
			{

				stream = connection.GetStream();
				//      formatter = new BinaryFormatter();
				connected = true;
			}
			catch (Exception e)
			{
				Console.WriteLine(e.StackTrace);
			}
		}

		public virtual void run()
		{
			while (connected)
			{
				try
				{
					Request request = Request.ParseDelimitedFrom(stream);
					Response response = handleRequest(request);
					if (response != null)
					{
						sendResponse(response);
					}
				}
				catch (Exception e)
				{
					Console.WriteLine(e.StackTrace);
				}

				try
				{
					Thread.Sleep(1000);
				}
				catch (Exception e)
				{
					Console.WriteLine(e.StackTrace);
				}
			}
			try
			{
				stream.Close();
				connection.Close();
			}
			catch (Exception e)
			{
				Console.WriteLine("Error " + e);
			}
		}
		/*public virtual void messageReceived(chat.model.Message message)
		{
			//MessageDTO mdto = DTOUtils.getDTO(message);
			Console.WriteLine("Message received  " + message);
			try
			{
				sendResponse(ProtoUtils.createNewMessageResponse(message));
			}
			catch (Exception e)
			{
				throw new ChatException("Sending error: " + e);
			}
		}

		public virtual void friendLoggedIn(chat.model.User friend)
		{
			//UserDTO udto =DTOUtils.getDTO(friend);
			Console.WriteLine("Friend logged in " + friend);
			try
			{
				sendResponse(ProtoUtils.createFriendLoggedInResponse(friend));
			}
			catch (Exception e)
			{
				Console.WriteLine(e.StackTrace);
			}
		}
		public virtual void friendLoggedOut(chat.model.User friend)
		{
			//UserDTO udto =DTOUtils.getDTO(friend);
			Console.WriteLine("Friend logged out " + friend);
			try
			{
				sendResponse(ProtoUtils.createFriendLoggedOutResponse(friend));
			}
			catch (Exception e)
			{
				Console.WriteLine(e.StackTrace);
			}
		}*/

		private Response handleRequest(Request request)
		{
			Response response = null;
			Request.Types.Type reqType = request.Type;
			switch (reqType)
			{
				case Request.Types.Type.Login:
					{
						Console.WriteLine("Login request ...");
						//LoginRequest logReq =(LoginRequest)request;
						//UserDTO udto =logReq.User;
						AppModel.Utilizator user = ProtoUtils.getUser(request);
						try
						{
							lock (server)
							{
								server.login(user, this);
							}
							return ProtoUtils.createOkResponse();
						}
						catch (AppException e)
						{
							connected = false;
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}
				case Request.Types.Type.Logout:
					{
						Console.WriteLine("Logout request");
						//LogoutRequest logReq =(LogoutRequest)request;
						//UserDTO udto =logReq.User;
						AppModel.Utilizator user = ProtoUtils.getUser(request);
						try
						{
							lock (server)
							{

								server.logout(user, this);
							}
							connected = false;
							return ProtoUtils.createOkResponse();

						}
						catch (AppException e)
						{
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}
				/*case Request.Types.Type.SendMessage:
					{
						Console.WriteLine("SendMessageRequest ...");
						//SendMessageRequest senReq =(SendMessageRequest)request;
						//MessageDTO mdto =senReq.Message;
						chat.model.Message message = ProtoUtils.getMessage(request);
						try
						{
							lock (server)
							{
								server.sendMessage(message);
							}
							return ProtoUtils.createOkResponse();
						}
						catch (ChatException e)
						{
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}

				case ChatRequest.Types.Type.GetLoggedFriends:
					{
						Console.WriteLine("GetLoggedFriends Request ...");
						//GetLoggedFriendsRequest getReq =(GetLoggedFriendsRequest)request;
						//UserDTO udto =getReq.User;
						chat.model.User user = ProtoUtils.getUser(request);  //DTOUtils.getFromDTO(udto);
						try
						{
							chat.model.User[] friends;
							lock (server)
							{

								friends = server.getLoggedFriends(user);
							}
							//UserDTO[] frDTO =DTOUtils.getDTO(friends);
							return ProtoUtils.createLoggedFriendsResponse(friends);
						}
						catch (ChatException e)
						{
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}*/
			}
			return response;
		}

		private void sendResponse(Response response)
		{
			Console.WriteLine("sending response " + response);
			//formatter.Serialize(stream, response);
			response.WriteDelimitedTo(stream);
			stream.Flush();

		}

        public void newInscriere()
        {
            throw new NotImplementedException();
        }

        public void update(IEnumerable<AppModel.Proba> list)
        {
            throw new NotImplementedException();
        }
    }
}

File: 4th Semester/Design and programming environments/Project Java Client-Server/Networking/src/main/java/protobuffprotocol/ProtoChatWorker.java
package protobuffprotocol;


import domain.Utilizator;
import services.AppException;
import services.IAppObserver;
import services.IService;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.List;


public class ProtoChatWorker implements Runnable, IAppObserver {
    private IService server;
     private Socket connection;

     private InputStream input;
     private OutputStream output;
     private volatile boolean connected;
     public ProtoChatWorker(IService server, Socket connection) {
         this.server = server;
         this.connection = connection;
         try{
             output=connection.getOutputStream() ;//new ObjectOutputStream(connection.getOutputStream());
             input=connection.getInputStream(); //new ObjectInputStream(connection.getInputStream());
             connected=true;
         } catch (IOException e) {
             e.printStackTrace();
         }
     }

     public void run() {
         while(connected){
             try {
                // Object request=input.readObject();
                 System.out.println("Waiting requests ...");
                 ChatProtobufs.Request request=ChatProtobufs.Request.parseDelimitedFrom(input);
                 System.out.println("Request received: "+request);
                 ChatProtobufs.Response response=handleRequest(request);
                 if (response!=null){
                    sendResponse(response);
                 }
             } catch (IOException e) {
                 e.printStackTrace();
             }
             try {
                 Thread.sleep(1000);
             } catch (InterruptedException e) {
                 e.printStackTrace();
             }
         }
         try {
             input.close();
             output.close();
             connection.close();
         } catch (IOException e) {
             System.out.println("Error "+e);
         }
     }

//     public void messageReceived(Message message) throws ChatException {
//         System.out.println("Message received  "+message);
//         try {
//             sendResponse(ProtoUtils.createNewMessageResponse(message));
//         } catch (IOException e) {
//             throw new ChatException("Sending error: "+e);
//         }
//     }
//
//     public void friendLoggedIn(User friend) throws ChatException {
//         System.out.println("Friend logged in "+friend);
//         try {
//             sendResponse(ProtoUtils.createFriendLoggedInResponse(friend));
//         } catch (IOException e) {
//             e.printStackTrace();
//         }
//     }
//
//     public void friendLoggedOut(User friend) throws ChatException {
//         System.out.println("Friend logged out "+friend);
//         try {
//             sendResponse(ProtoUtils.createFriendLoggedOutResponse(friend));
//         } catch (IOException e) {
//             e.printStackTrace();
//         }
//     }

     private ChatProtobufs.Response handleRequest(ChatProtobufs.Request request){
         ChatProtobufs.Response response=null;
         switch (request.getType()){
             case Login:{
                 System.out.println("Login request ...");
                 Utilizator user=ProtoUtils.getUser(request);
                 try {
                     server.login(user, this);
                     return ProtoUtils.createOkResponse();
                 } catch (AppException e) {
                     connected=false;
                     return ProtoUtils.createErrorResponse(e.getMessage());
                 }
             }
             case Logout:{
                 System.out.println("Logout request");
                 Utilizator user=ProtoUtils.getUser(request);
                 try {
                     server.logout(user, this);
                     connected=false;
                     return ProtoUtils.createOkResponse();

                 } catch (AppException e) {
                     return ProtoUtils.createErrorResponse(e.getMessage());
                 }
             }
//             case SendMessage:{
//                 System.out.println("SendMessageRequest ...");
//                 Message message=ProtoUtils.getMessage(request);
//                 try {
//                     server.sendMessage(message);
//                     return ProtoUtils.createOkResponse();
//                 } catch (ChatException e) {
//                     return ProtoUtils.createErrorResponse(e.getMessage());
//                 }
//             }
//             case GetLoggedFriends:{
//                 System.out.println("GetLoggedFriends Request ...");
//                 User user=ProtoUtils.getUser(request);
//                 try {
//                     User[] friends=server.getLoggedFriends(user);
//                     return ProtoUtils.createLoggedFriendsResponse(friends);
//                 } catch (ChatException e) {
//                     return ProtoUtils.createErrorResponse(e.getMessage());
//                 }
//             }
         }
         return response;
     }

     private void sendResponse(ChatProtobufs.Response response) throws IOException{
         System.out.println("sending response "+response);
         response.writeDelimitedTo(output);
         //output.writeObject(response);
         output.flush();
     }

    @Override
    public void newInscriere() throws AppException {

    }
}
